man/diodmount.8.in
.TH mount.diod 8 "@PACKAGE_VERSION@" "@PACKAGE_NAME@" "@PACKAGE_NAME@"
.SH NAME
mount.diod \- mount diod file systems
.SH SYNOPSIS
\fBmount.diod\fR \fI[OPTIONS] host[:aname] dir [-o options]\fR
.SH DESCRIPTION
.B mount.diod
is the mount utility for the distributed I/O daemon, which is an I/O
forwarding service for Linux clusters built upon the Linux \fB9p\fR
file system.
.LP
\fBmount.diod\fR performs two main tasks:
1) connect and authenticate to the \fBdiod\fR server, and
2) issue the \fImount(2)\fR system call with appropriate mount options.
.LP
Authentication with the \fBdiod\fR server takes place in user space.
The connected (and authenticated) file descriptor is passed as a mount
option to the kernel \fB9p\fR file system
.LP
This command requires root to execute the \fImount(2)\fR system call.
.LP
The \fB9p\fR file system must support the 9P2000.L protocol variant,
e.g. kernel 2.6.38 or above.
.LP
The \fIhost\fR portion of the mount spec may optionally be supplied in
hostlist format, e.g. host1,host2,host3:aname or host[1-3]:aname, in which
case hosts are tried in order until a successful connection is obtained.
Post-connect mount errors are immediately fatal.
.LP
If the \fIhost\fR portion of the mount spec begins with a forward-slash (/),
it refers to the path to a UNIX domain socket.
.LP
IPv6 addresses may be used for the \fIhost\fR portion of the mount spec if they
are enclosed in square brackets.
.SH COMMAND LINE OPTIONS
.TP
.I "-f, --fake-mount"
Do everything except the mount system call.
.TP
.I "-n, --no-mtab"
Do not update the system mtab.
.TP
.I "-v, --verbose"
Set verbose mode.
.TP
.I "-o, --options opt[,opt,...]"
Set mount options (see below).
.TP
.I "-p, --privport"
Connect from a socket bound to a port in the range of 512-1023,
available to root only.  This can be used in conjunction with the
\fIprivport\fR export option.
.SH MOUNT OPTIONS
The following file system specific mount options are handled by
\fBmount.diod\fR.
.TP
.I "aname=name"
The aname is the path of the exported file system on the server.
It can be specified in the mount spec in \fIhost:aname\fR format
like NFS, or it can be specified with this option.
\fBdiod\fR will not allow an attach to an empty aname.
\fBdiod\fR has a built-in synthetic control file system that can be
mounted with aname of \fIctl\fR.
Note that "ctl" must be explicitly exported.
.TP
.I "uname=USERNAME"
The user name to authenticate to the \fBdiod\fR server.
If neither \fIuname\fR nor \fIaccess\fR are specified,
\fIuname=root,access=user\fR is assumed.
.TP
.I "access=user"
The client allows any user to access the file system and enforces traditional
UNIX access controls.  Each new user attaches to the server.
Requires \fIuname=root\fR.
.TP
.I "access=client,posixacl"
The client allows any user to access the file system and enforces POSIX ACLs.
Only the authenticated user attaches to the server.
Requires \fIuname=root\fR.
.TP
.I "access=UID"
The client only allows UID to access the filesystem.
Only the authenticated user attaches to the server.
.TP
.I "access=any"
The client allows any user to access the filesystem.
Only the authenticated user attaches to the server.
.TP
.I "port=n"
Connect to the \fIdiod\fR server on the specified port.
The default is the IANA-assigned 9pfs port 564.
.TP
.I "trans=name"
Select a transport.
\fBmount.diod\fR only supports \fItrans=fd\fR
because of its strategy for authentication.
.TP
.I "rfdno=n,wfdno=n"
The file descriptor for reading and writing with \fItrans=fd\fR.
If set, \fBmount.diod\fR assumes this file descriptor is already
connected to a \fBdiod\fR server and ignores \fIhost\fR in the
device specification, and the \fIport=n\fR option.
.TP
.I "msize=n"
The msize is the max 9P packet payload size.
The default is 65512 (65536-24)
.TP
.I "version=name"
Select 9P protocol version.
\fBdiod\fR only supports \fI9p2000.L\fR.
.TP
.I "cache=mode"
Specify a caching policy.  By default, no caches are used.
If \fIcache=loose\fR, no attempts are made at consistency.
This mode is intended for exclusive, read-only mounts.
If \fIcache=fscache\fR, use FS-Cache for a persistent, read-only
cache backend.
.TP
.I "cachetag"
In the context of \fIcache=fscache\fR, select the cache tag
to use for this session.
Cache tags for existing cache sessions are listed in /sys/fs/9p/caches.
.TP
.I "nodevmap"
Do not map special files.  Represent them as normal files.
This can be used to share devices/named pipes/sockets between
hosts.
.TP
.I "debug=n"
Specifies debug level for the kernel 9p module.  The debug level is a bitmask.
.br
0x01  = display verbose error messages
.br
0x02  = developer debug (DEBUG_CURRENT)
.br
0x04  = display 9p trace
.br
0x08  = display VFS trace
.br
0x10  = display Marshalling debug
.br
0x20  = display RPC debug
.br
0x40  = display transport debug
.br
0x80  = display allocation debug
.br
0x100 = display protocol message debug
.br
0x200 = display Fid debug
.br
0x400 = display packet debug
.br
0x800 = display fscache tracing debug
.br
The debug level is global, i.e. it applies to all 9P file systems and
it remains set after the file system is unmounted.
The default is \fIdebug=0\fR.
.SH "SEE ALSO"
diod (8)


NEWS
=========================================================================
Release Notes for DIOD version 1.0.24                         31 Mar 2015
=========================================================================

* Replace sysv-init script with systemd unit file

* Move fully to github (including docs)

* Misc. minor updates, mostly tests and build environment, for rhel7

=========================================================================
Release Notes for DIOD version 1.0.23                         26 Aug 2014
=========================================================================

* Add dist tag to release (TOSS-2662)

* Minor autotools forward-portability (issue 124)

* Doc: AUTHORS, README, license headers updated

* Stop tracking additional autotools build products.

=========================================================================
Release Notes for DIOD version 1.0.22                         19 Aug 2014
=========================================================================

* Stop tracking autotools build products in git.

* Fix assertion failure at srv.c:754 (issue 122)

* Portability fixes for musl libc (Felix Janda)

* Change default rdma port to 5640 to match kernel

* rdmatrans: only register the used buffer to the network
  adapter (Dominique Martinet)

* fix configure --with-tcmalloc logic

* libnpfs: fids cloned in xattrwalk should inherit flags

* rdmatrans: compilation error

=========================================================================
Release Notes for DIOD version 1.0.21                         31 May 2013
=========================================================================

* Require that "ctl" be explicitly exported.  This allows export options
  such as "ro", "noauth", and "privport" to apply to the ctl synthetic
  file system.

* The "auth_required_ctl" config option introduced in 1.0.15 is now
  deprecated.

* Add test coverage for trusted.* and security.* xattr namespaces.

* Add more mmap test coverage.

=========================================================================
Release Notes for DIOD version 1.0.20                         17 May 2013
=========================================================================

* Fix diodshowmount segfault on exit with 'noauth'.

* Add diodls --privport option.

* Additional test coverage (atomic create test, pathwalk test)

=========================================================================
Release Notes for DIOD version 1.0.19                         07 May 2013
=========================================================================

* Additional test coverage for xattrs:
  - Verify that setxattr flags XATTR_CREATE, XATTR_REPLACE work.
  - Verify that that removal of a nonexistent user xattr returns ENODATA

* Fix failure of tests/user/t17 when SELinux is enabled (bug in the test)

=========================================================================
Release Notes for DIOD version 1.0.18                         06 May 2013
=========================================================================

* Man page update.

=========================================================================
Release Notes for DIOD version 1.0.17                         06 May 2013
=========================================================================

* RPM: add missing BuildRequires for 'libattr-devel'.

=========================================================================
Release Notes for DIOD version 1.0.15                         03 May 2013
=========================================================================

* Add xattr support (issue 95)

* Support for dracut bootstrap of 9nbd network block device:
  - New 'dioddate' utility to synchronize system time with server's,
    and 'auth_required_ctl' config option which allows MUNGE
    authentication to be disabled for the ctl synthetic file system.
  - Add 'noauth' and 'privport' export options which disable MUNGE auth,
    and require attaches to originate from a privileged port (512-1024)
    for a specific export.    

=========================================================================
Release Notes for DIOD version 1.0.14                         12 Jul 2012
=========================================================================

* Add mount.diod --attach and --detach options to support 9nbd
  (9P network block device).

* Allow non-directories to be exported (issue 113)

* Return Rlerror(ENOENT) in response to Tauth if auth not required
  instead of Rlerror(0), which confused kernel 9P code.
    
=========================================================================
Release Notes for DIOD version 1.0.13                         22 May 2012
=========================================================================

* Allow DIOD_MOUNTOPTS in /etc/sysconfig/auto.diod

  Normalize -s,--server NAME option in diodcat, diodload, diodls,
  and diodshowmount utilities, where NAME can be IP[:PORT], HOST[:PORT],
  or /path/to/socket.

* Add diodshowmount utility (issue 111)

* Add diodls utility

* Debugging/cleanup for issue 108
  - refactor and clarify error paths in ioctx open/close functions
  - normalize error messages when ioctx state is not as expected

=========================================================================
Release Notes for DIOD version 1.0.12                         15 May 2012
=========================================================================

* Rework fdtrans somewhat for clarity (issue 109)

* Test improvements:
  - misc/t15: don't use munge auth or usleep (caused hang on build farm)
  - misc/t15: update valgrind suppressions for 32b Ubuntu tests

=========================================================================
Release Notes for DIOD version 1.0.11                         14 May 2012
=========================================================================

* Return EINVAL when a file is opened with O_DIRECT flag (issue 110).

* Drop unused maxmmap feature.


src/cmd/diodmount.c
/************************************************************\
 * Copyright 2010 Lawrence Livermore National Security, LLC
 * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
 *
 * This file is part of the diod 9P server project.
 * For details, see https://github.com/chaos/diod.
 *
 * SPDX-License-Identifier: GPL-2.0-or-later
\************************************************************/

/* diodmount.c - diod file system mount helper
 *
 * Usage: /sbin/mount.diod spec dir [-sfnv] [-o options]
 */

#if HAVE_CONFIG_H
#include "config.h"
#endif
#include <sched.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdlib.h>
#include <sys/socket.h>
#include <sys/wait.h>
#include <stdint.h>
#include <netdb.h>
#include <stdio.h>
#include <getopt.h>
#include <mntent.h>
#include <sys/mount.h>
#include <string.h>
#include <errno.h>
#include <ctype.h>
#include <pwd.h>
#include <libgen.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <paths.h>


#ifndef PATH_MAX
#define PATH_MAX 1024
#endif

#include "src/libnpfs/npfs.h"
#include "src/libnpclient/npclient.h"
#include "src/liblsd/list.h"
#include "src/liblsd/hostlist.h"
#include "src/libdiod/diod_log.h"
#include "src/libdiod/diod_sock.h"
#include "src/libdiod/diod_auth.h"
#include "opt.h"

static const char *options = "fnvo:p";

static const struct option longopts[] = {
    {"fake-mount",      no_argument,         0, 'f'},
    {"no-mtab",         no_argument,         0, 'n'},
    {"verbose",         no_argument,         0, 'v'},
    {"options",         required_argument,   0, 'o'},
    {"privport",        no_argument,         0, 'p'},
    {0, 0, 0, 0},
};

#define NBD_SET_BLKSIZE _IO( 0xab, 1 )
#define NBD_SET_TIMEOUT _IO( 0xab, 2 )
#define NBD_START       _IO( 0xab, 3 )
#define NBD_STOP        _IO( 0xab, 4 )

#define NBD_SET_OPTS    _IOW( 0xab, 10, char* )
#define NBD_SET_ADDR    _IOW( 0xab, 11, char* )
#define NBD_SET_PATH    _IOW( 0xab, 12, char* )

#define DIOD_DEFAULT_MSIZE 65536
static uid_t _uname2uid (char *uname);
static void _diod_mount (Opt o, int rfd, int wfd, char *spec, char *dir,
                         int vopt, int fopt, int nopt);
static void _diod_remount (Opt o, char *spec, char *dir, int vopt, int fopt);
static void _verify_mountpoint (char *path);
static void _parse_uname_access (Opt o);
static char *_parse_spec (char *spec, Opt o);
static const char *_get_spec_suffix (const char *spec);
static void _mount (const char *source, const char *target,
                    unsigned long mountflags, const void *data);

static void
usage (void)
{
    fprintf (stderr,
"Usage: mount.diod [OPTIONS] host[:aname] directory\n"
"   -f,--fake-mount               do everything but the actual mount\n"
"   -n,--no-mtab                  do not update /etc/mtab\n"
"   -v,--verbose                  verbose mode\n"
"   -o,--options opt[,opt,...]    specify mount options\n"
"   -p,--privport                 bind localy to a privileged port\n"
);
    exit (1);
}

int
main (int argc, char *argv[])
{
    char *dir = NULL;
    char *spec, *host;
    char *nspec = NULL;
    int c;
    int nopt = 0;
    int vopt = 0;
    int fopt = 0;
    int privport = 0;
    int rfd = -1, wfd = -1;
    Opt o;

    diod_log_init (argv[0]);

    o = opt_create ();

    opterr = 0;
    while ((c = getopt_long (argc, argv, options, longopts, NULL)) != -1) {
        switch (c) {
            case 'f':   /* --fake-mount */
                fopt = 1;
                break;
            case 'n':   /* --no-mtab */
                nopt = 1;
                break;
            case 'v':   /* --verbose */
                vopt++;
                break;
            case 'o':   /* --options OPT[,OPT]... */
                opt_addf (o, "%s", optarg);
                break;
            case 'p':   /* --privport */
                privport++;
                break;
            default:
                usage ();
        }
    }

    if (optind != argc - 2)
        usage ();

    if (geteuid () != 0)
        msg_exit ("you must be root");

    spec = argv[optind++];
    dir = argv[optind++];
    host = _parse_spec (spec, o);

    _verify_mountpoint (dir);

    /* Remount - only pass mount flags into the VFS for an existing mount.
     * Take care of it here and exit.
     */
    if (opt_find (o, "remount")) {
        if (opt_check_allowed_csv (o, "ro,rw,aname,remount"))
            msg_exit ("-oremount can only be used with ro,rw");
        _diod_remount (o, spec, dir, vopt, fopt);
        goto done;
    }

    /* Ensure uname and access are set, and to diod-compatible values.
     * The uname user becomes the euid which will be used by munge auth.
     */
    _parse_uname_access (o);
     if (seteuid (_uname2uid (opt_find (o, "uname"))) < 0)
        err_exit ("seteuid");

    /* We require -otrans=fd because auth occurs in user space, then live fd
     * is passed to the kernel via -orfdno,wfdno.
     */
    if (!opt_find (o, "trans"))
        opt_addf (o, "trans=%s", "fd");
    else if (!opt_find (o, "trans=fd"))
        msg_exit ("only -otrans=fd transport is supported");

    /* Set msize if not already set.  Validate it later.
     */
    if (!opt_find (o, "msize"))
        opt_addf (o, "msize=%d", DIOD_DEFAULT_MSIZE);

    /* Only .L version is supported.
     */
    if (!opt_find (o, "version"))
        opt_addf (o, "version=%s", "9p2000.L");
    else if (!opt_find (o, "version=9p2000.L"))
        msg_exit ("only -oversion=9p2000.L is supported (little p, big L)");

    /* Set debug level.
     */
    if (!opt_find (o, "debug"))
        opt_addf (o, "debug=%d", 0x1); /* send errors to dmesg */

    /* Set rwdepth (number of concurrent reads with buffer > msize).
     * N.B. this option is not upstream yet but unknown options are ignored.
     */
    if (!opt_find (o, "rwdepth"))
        opt_addf (o, "rwdepth=%d", 1);

    /* Server is on an inherited file descriptor.
     * For testing, we start server on a socketpair duped to fd 0.
     */
    if (opt_find (o, "rfdno") || opt_find (o, "wfdno")) {
        if (!opt_scanf (o, "rfdno=%d", &rfd) || !opt_scanf (o, "wfdno=%d",&wfd))
            msg_exit ("-orfdno,wfdno must be used together");
        nopt = 1; /* force no mtab */

    /* Connect to server on UNIX domain socket
     */
    } else if (host[0] == '/') {
        if (opt_find (o, "port"))
            msg_exit ("-oport won't work with UNIX domain socket");
        if ((rfd = diod_sock_connect_unix (host, 0)) < 0)
            exit (1);
        wfd = rfd;

        opt_addf (o, "rfdno=%d", rfd);
        opt_addf (o, "wfdno=%d", wfd);

    /* Connect to server on IANA port (or user-specified) and host.
     */
    } else {
        char *port = opt_find (o, "port");
        hostlist_iterator_t hi;
        hostlist_t hl;
        char *h;

        if (!port)
            port = "564";
        if (!(hl = hostlist_create (host)))
            msg_exit ("error parsing host string: %s", host);
        if (!(hi = hostlist_iterator_create (hl)))
            msg_exit ("out of memory");
        while ((h = hostlist_next (hi))) {
            int flags = DIOD_SOCK_QUIET;
            if (privport)
                flags |= DIOD_SOCK_PRIVPORT;
            if (vopt)
                msg ("trying to connect to %s:%s", h, port);
            if ((rfd = diod_sock_connect_inet (h, port, flags )) >= 0)
                break;
        }
        if (h) { /* create new 'spec' string identifying successful host */
            const char *p = _get_spec_suffix (spec);
            int len = strlen (h) + (p ? strlen (p) : 0) + 1;

            if (!(nspec = malloc (len)))
                msg_exit ("out of memory");
            snprintf (nspec, len, "%s%s", h, p ? p : "");
        }
        hostlist_destroy (hl);
        if (rfd < 0)
            msg_exit ("could not connect to server(s), giving up");
        wfd = rfd;

        opt_delete (o, "port");
        opt_addf (o, "rfdno=%d", rfd);
        opt_addf (o, "wfdno=%d", wfd);
    }

    /* At this point options (o) will include:
     *   trans=fd,rfdno=N,wfdno=N,msize=N,version=9p2000.L,debug=N,
     *   uname=USERNAME,access=MODE
     */
    _diod_mount (o, rfd, wfd, nspec ? nspec : spec, dir, vopt, fopt, nopt);

done:
    opt_destroy (o);
    exit (0);
}

static uid_t
_uname2uid (char *uname)
{
    struct passwd *pw;

    if (!(pw = getpwnam (uname)))
        msg_exit ("could not look up uname='%s'", uname);
    return pw->pw_uid;
}

/* Parse HOST[:ANAME] specification.
 * If ANAME is specified, add -oaname=ANAME to mount options.
 * Returns a heap copy of host.
 * Exits on error.
 */
static char *
_parse_spec (char *spec, Opt o)
{
    char *host, *hostend, *aname;
    int ipv6 = 0;

    if (spec[0] == '[') {
        ipv6 = 1;
        spec++;
    }

    if (!(host = strdup (spec)))
        msg_exit ("out of memory");
    if (ipv6) {
        hostend = strchr(host, ']');
        if (!hostend)
            msg_exit ("invalid IPv6 address");
        *hostend++ = '\0';
    } else {
        hostend = host;
    }
    if ((aname = strchr (hostend, ':')))
        *aname++ = '\0';
    if (strlen (host) == 0)
        msg_exit ("no host specified");
    if (!aname || strlen (aname) == 0)
        ; /* aname = opt_find (o, "aname"); */
    else if (!opt_addf (o, "aname=%s", aname))
        msg_exit ("you cannot have both -oaname and spec=host:aname");

    return host;
}

static const char *
_get_spec_suffix (const char *spec)
{
    if (spec[0] == '[') {
        spec = strchr(spec, ']');
        NP_ASSERT (spec);
    }

    return strchr(spec, ':');
}

/* Verify that directory exists, exiting on failure.
 */
static void
_verify_mountpoint (char *path)
{
    struct stat sb;

    if (stat (path, &sb) < 0)
        err_exit ("stat %s", path);
    if (!S_ISDIR (sb.st_mode))
        msg_exit ("%s: not a directory", path);
}

/* Provide defaults for uname and access mount options.
 * If either are unset, abort with helpful hints.
 */
static void
_parse_uname_access (Opt o)
{
    char *uname = opt_find (o, "uname");
    char *access = opt_find (o, "access");

    if (!uname && !access) {
        opt_addf (o, "uname=%s", "root");
        opt_addf (o, "access=%s", "user");
    }
    else if (!uname || !access) {
        msg_exit (
"access,uname mount options must be set\n"
"Common examples:\n"
"  -o uname=root,access=user            Multi-user with UNIX access controls\n"
"  -o uname=root,access=client,posixacl Multi-user with POSIX ACLs\n"
"  -o uname=USERNAME,access=UID         Single-user access");
    }
}

typedef struct {
    char *opt;
    unsigned long flag;
} map_t;

static map_t setopt[] = {
    { "noatime",        MS_NOATIME},
    { "nodev",          MS_NODEV},
    { "nodiratime",     MS_NODIRATIME},
    { "noexec",         MS_NOEXEC},
    { "nosuid",         MS_NOSUID},
    { "ro",             MS_RDONLY },
    { "remount",        MS_REMOUNT},
    { "sync",           MS_SYNCHRONOUS},
};

static map_t clropt[] = {
    { "atime",          MS_NOATIME},
    { "dev",            MS_NODEV},
    { "diratime",       MS_NODIRATIME},
    { "exec",           MS_NOEXEC},
    { "suid",           MS_NOSUID},
    { "rw",             MS_RDONLY },
    { "nosync",         MS_SYNCHRONOUS},
};

static void
_getflags (Opt o, unsigned long *flags)
{
    int i;

    for (i = 0; i < sizeof (setopt) / sizeof (map_t); i++) {
        if (opt_find (o, setopt[i].opt)) {
            *flags |= setopt[i].flag;
            opt_delete (o, setopt[i].opt);
        }
    }
    for (i = 0; i < sizeof (clropt) / sizeof (map_t); i++) {
        if (opt_find (o, clropt[i].opt)) {
            *flags &= ~clropt[i].flag;
            opt_delete (o, clropt[i].opt);
        }
    }
}

static void
_diod_remount (Opt o, char *spec, char *dir, int vopt, int fopt)
{
    char *options = opt_csv (o);
    unsigned long mountflags = 0;

    _getflags (o, &mountflags);
    if (vopt)
        msg ("mount %s %s -o%s", spec, dir, options);
    if (!fopt)
        _mount (spec, dir, mountflags, NULL);

    free (options);
}

static int
_update_mtab (char *options, char *spec, char *dir)
{
    uid_t saved_euid = geteuid ();
    FILE *f;
    int ret = 0;
    struct stat st;
    struct mntent mnt;

    ret = lstat(_PATH_MOUNTED, &st);

    /* /etc/mtab can be a symlink to /proc/mounts or
     * /proc/self/mounts, in which case we shouldn't write to it even
     * if /proc isn't mounted.  Return success in that case to avoid a
     * spurious error message. */
    if (ret != 0)
	return 0;

    if (!S_ISREG(st.st_mode))
	return 1;

    mnt.mnt_fsname = spec;
    mnt.mnt_dir = dir;
    mnt.mnt_type = "diod";
    mnt.mnt_opts = options;
    mnt.mnt_freq = 0;
    mnt.mnt_passno = 0;

    if (seteuid (0) < 0) {
        err ("failed to set effective uid to root");
        goto done;
    }
    if (!(f = setmntent (_PATH_MOUNTED, "a"))) {
        err (_PATH_MOUNTED);
        goto done;
    }
    if (addmntent (f, &mnt) != 0) {
        msg ("failed to add entry to %s", _PATH_MOUNTED);
        endmntent (f);
        goto done;
    }
    endmntent (f);
    if (seteuid (saved_euid) < 0) {
        err ("failed to restore effective uid to %d", saved_euid);
        goto done;
    }
    ret = 1;
done:
    return ret;
}

static void
_diod_mount (Opt o, int rfd, int wfd, char *spec, char *dir, int vopt,
             int fopt, int nopt)
{
    char *options, *options9p, *aname, *uname;
    uid_t uid;
    int msize;
    Npcfsys *fs;
    Npcfid *afid, *root;
    unsigned long mountflags = 0;

    options = opt_csv (o);
    _getflags (o, &mountflags);
    options9p = opt_csv (o); /* after mountflags removed from opt list */

    if (!(uname = opt_find (o, "uname")))
        msg_exit ("uname is not set"); /* can't happen */
    uid = _uname2uid (uname);
    aname = opt_find (o, "aname"); /* can be null */
    if (!opt_scanf (o, "msize=%d", &msize) || msize < IOHDRSZ)
        msg_exit ("msize must be set to integer >= %d", IOHDRSZ);

    if (vopt)
        msg ("pre-authenticating connection to server");
    if (!(fs = npc_start (rfd, wfd, msize, 0)))
        errn_exit (np_rerror (), "version");
    if (!(afid = npc_auth (fs, aname, uid, diod_auth)) && np_rerror () != 0)
        errn_exit (np_rerror (), "auth");
    if (!(root = npc_attach (fs, afid, aname, uid))) {
        errn (np_rerror (), "attach");
        if (afid)
            (void)npc_clunk (afid);
        exit (1);
    }
    if (afid && npc_clunk (afid) < 0)
        errn_exit (np_rerror (), "clunk afid");
    if (npc_clunk (root) < 0)
        errn_exit (np_rerror (), "clunk root");
    if (vopt)
        msg ("mount -t 9p %s %s -o%s", spec, dir, options);
    if (!fopt) {
        /* kernel wants non-blocking */
        if (fcntl (rfd, F_SETFL, O_NONBLOCK) < 0)
            err_exit ("setting O_NONBLOCK flag on rfd=%d", rfd);
        if (fcntl (wfd, F_SETFL, O_NONBLOCK) < 0)
            err_exit ("setting O_NONBLOCK flag on wfd=%d", wfd);
        _mount (spec, dir, mountflags, options9p);
    }
    npc_finish (fs); /* closes fd */

    if (!nopt) {
        if (!_update_mtab (options, spec, dir))
            msg_exit ("failed to update /etc/mtab");
    }
    free (options);
    free (options9p);
}

/* Mount 9p file system [source] on [target] with options [data].
 * Swap effective (user) and real (root) uid's for the duration of mount call.
 * Exit on error.
 */
static void
_mount (const char *source, const char *target, unsigned long mountflags,
        const void *data)
{
    uid_t saved_euid = geteuid ();

    if (seteuid (0) < 0)
        err_exit ("failed to set effective uid to root");
    if (mount (source, target, "9p", mountflags, data) < 0)
        err_exit ("mount");
    if (seteuid (saved_euid) < 0)
        err_exit ("failed to restore effective uid to %d", saved_euid);
}

/*
 * vi:tabstop=4 shiftwidth=4 expandtab
 */
